src/core/src/main/java/org/locationtech/geogig/storage/cache/SharedCache.java

```diff
@@ -0,0 +1,329 @@
+    };
+
+    public static SharedCache build(long maxCacheSizeBytes) {
+        return SharedCache.build(Impl.L1_CACHE_SIZE, maxCacheSizeBytes);
+    }
+
+    static SharedCache build(int L1capacity, long maxCacheSizeBytes) {
+        checkArgument(L1capacity >= 0);
+        checkArgument(maxCacheSizeBytes >= 0, "Cache size can't be < 0, 0 meaning no cache at all");
+
+        if (0L == maxCacheSizeBytes) {
+            return NO_CACHE;
+        }
+
+        int initialCapacityCount = 1_000_000;
+        int concurrencyLevel = 16;
+
+        CacheBuilder<Object, Object> cacheBuilder = CacheBuilder.newBuilder();
+        cacheBuilder = cacheBuilder.maximumWeight(maxCacheSizeBytes);
+        cacheBuilder.weigher(Impl.WEIGHER);
+
+        cacheBuilder.initialCapacity(initialCapacityCount);
+        cacheBuilder.concurrencyLevel(concurrencyLevel);
+        cacheBuilder.recordStats();
+
+        Impl.SizeTracker sizeTracker = new Impl.SizeTracker();
+        cacheBuilder.removalListener(sizeTracker);
+
+        Cache<Key, byte[]> byteCache = cacheBuilder.build();
+
+        return new Impl(L1capacity, byteCache, sizeTracker);
+    }
+
+    static final CacheStats EMPTY_STATS = new CacheStats(0, 0, 0, 0, 0, 0);
+
+    default boolean contains(Key id) {
+        return false;
+    }
+
+    default void invalidateAll() {
+    }
+
+    default void invalidateAll(CacheIdentifier prefix) {
+    }
+
+    default void dispose() {
+    }
+
+    default void invalidate(Key id) {
+    }
+
+    default @Nullable RevObject getIfPresent(Key key) {
+        return null;
+    }
+
+    default @Nullable Future<?> put(Key key, RevObject obj) {
+        return null;
+    }
+
+    default long sizeBytes() {
+        return 0L;
+    }
+
+    default long objectCount() {
+        return 0L;
+    }
+
+    default CacheStats getStats() {
+        return EMPTY_STATS;
+    }
+
+    static class Impl implements SharedCache {
+        /**
+         * 
+         */
+        static final ExecutorService WRITE_BACK_EXECUTOR;
+        static {
+            ThreadFactory tf = new ThreadFactoryBuilder().setDaemon(true)
+                    .setNameFormat("GeoGig shared cache %d").build();
+
+            final int nThreads = Math.max(2, Runtime.getRuntime().availableProcessors());
+
+            RejectedExecutionHandler sameThreadHandler = new ThreadPoolExecutor.CallerRunsPolicy();
+
+            WRITE_BACK_EXECUTOR = new ThreadPoolExecutor(1, nThreads, 30L, TimeUnit.SECONDS,
+                    new ArrayBlockingQueue<Runnable>(nThreads), tf, sameThreadHandler);
+        }
+
+        private static final ObjectSerializingFactory ENCODER = //
+                new LZ4SerializationFactory(//
+                        DataStreamSerializationFactoryV2_3.INSTANCE//
+                );
+
+        /**
+         * Size of the L1 cache {@link Key} -> {@link RevObject}
+         */
+        private static final int L1_CACHE_SIZE = 10_000;
```

Comment:
I'm just worried about using a bit too much memory - 10,000 * 10kb = 100Mb, so maybe it will usually not be a big deal.  But, if had large features, 10,000 * 100kb = 1GB, which might be a bit too much.